tests/handlers/oidc_test.py

```diff
@@ -0,0 +1,506 @@
+    }
+    assert data["access_token"] == data["id_token"]
+    verifier = setup.factory.create_token_verifier()
+    token = verifier.verify_internal_token(Token(encoded=data["id_token"]))
+    assert token.claims == {
+        "act": {
+            "aud": setup.config.issuer.aud,
+            "iss": setup.config.issuer.iss,
+            "jti": ANY,
+        },
+        "aud": setup.config.issuer.aud_internal,
+        "email": "githubuser@example.com",
+        "exp": ANY,
+        "iat": ANY,
+        "iss": setup.config.issuer.iss,
+        "jti": SessionHandle.from_str(code).key,
+        "name": "GitHub User",
+        "scope": "openid",
+        "sub": "githubuser",
+        "uid": "githubuser",
+        "uidNumber": "123456",
+    }
+    now = time.time()
+    expected_exp = now + setup.config.issuer.exp_minutes * 60
+    assert expected_exp - 5 <= token.claims["exp"] <= expected_exp
+    assert now - 5 <= token.claims["iat"] <= now
+
+    log = json.loads(caplog.record_tuples[0][2])
+    assert log == {
+        "event": "Retrieved token for user githubuser via OpenID Connect",
+        "level": "info",
+        "logger": "gafaelfawr",
+        "method": "POST",
+        "path": "/auth/openid/token",
+        "remote": "127.0.0.1",
+        "request_id": ANY,
+        "scope": "openid",
+        "token": SessionHandle.from_str(code).key,
+        "user": "githubuser",
+        "user_agent": ANY,
+    }
+
+
+async def test_unauthenticated(
+    create_test_setup: SetupTestCallable, caplog: LogCaptureFixture
+) -> None:
+    clients = [OIDCClient(client_id="some-id", client_secret="some-secret")]
+    setup = await create_test_setup(oidc_clients=clients)
+
+    return_url = f"https://{setup.client.host}:4444/foo?a=bar&b=baz"
+    login_params = {
+        "response_type": "code",
+        "scope": "openid",
+        "client_id": "some-id",
+        "state": "random-state",
+        "redirect_uri": return_url,
+    }
+    caplog.clear()
+    r = await setup.client.get(
+        "/auth/openid/login", params=login_params, allow_redirects=False,
+    )
+
+    assert r.status == 302
+    url = urlparse(r.headers["Location"])
+    assert not url.scheme
+    assert not url.netloc
+    assert url.path == "/login"
+    expected_url = setup.client.make_url("/auth/openid/login").with_query(
+        **login_params
+    )
+    assert query_from_url(r.headers["Location"]) == {"rd": [str(expected_url)]}
+
+    log = json.loads(caplog.record_tuples[0][2])
+    assert log == {
+        "event": "Redirecting user for authentication",
+        "level": "info",
+        "logger": "gafaelfawr",
+        "method": "GET",
+        "path": "/auth/openid/login",
+        "remote": "127.0.0.1",
+        "request_id": ANY,
+        "user_agent": ANY,
+    }
+
+
+async def test_login_errors(
+    create_test_setup: SetupTestCallable, caplog: LogCaptureFixture
+) -> None:
+    clients = [OIDCClient(client_id="some-id", client_secret="some-secret")]
```

Comment:
since you use some-id and some-secret a lot, maybe just make it a constant?  that way it's a bit easier to see where you use it in the code, since strings don't stick out as much

Reply:
Or maybe just pull them out?  I'm not sure where you override it

Reply:
Hm, they're just arbitrary strings that we can check match in other places later on. There are no default values, so it's not overridden; if the test doesn't set them, there would be no valid clients. They feel the same as strings like `"githubuser"` or `"random-state"`; they're data that we're testing is persisted or checked properly through the code.

But I'm not sure if that's what you're asking, or if I'm correctly envisioning the alternative you're thinking of? (I agree that this test is long and tedious to read, and it's not totally clear what's being checked, but I wasn't sure how to make it better. It's a fairly mechanical exploration of the error paths.)

Reply:
Ah I gotcha that makes sense, yeah it is a bit mechanical, but I guess that's fine